sbt-version-policy/src/main/scala/sbtversionpolicy/DependencyCheckReport.scala

```diff
@@ -61,44 +62,120 @@ object DependencyCheckReport {
               case (matcher, rec) if matcher.matches(org, name) =>
                 rec
             }
             .getOrElse(defaultReconciliation)
-          if (reconciliation.isCompatible(ver, currentVersion))
-            CompatibleVersion(currentVersion, ver, reconciliation)
+          val isCompatible =
+            if (compatibilityIntention == Compatibility.BinaryAndSourceCompatible) {
+              isBinaryCompatible(currentVersion, previousVersion, reconciliation) &&
+                isSourceCompatible(currentVersion, previousVersion, reconciliation)
+            } else {
+              isBinaryCompatible(currentVersion, previousVersion, reconciliation)
+            }
+          if (isCompatible)
+            CompatibleVersion(currentVersion, previousVersion, reconciliation)
           else
-            IncompatibleVersion(currentVersion, ver, reconciliation)
+            IncompatibleVersion(currentVersion, previousVersion, reconciliation)
       }
 
-      orgName -> status
+      orgAndName -> status
+    }
+
+  private def isBinaryCompatible(currentVersion: String, previousVersion: String, versionCompatibility: VersionCompatibility): Boolean =
+    versionCompatibility.isCompatible(previousVersion, currentVersion)
+
+  private def isSourceCompatible(currentVersion: String, previousVersion: String, versionCompatibility: VersionCompatibility): Boolean =
+    versionCompatibility match {
+      case VersionCompatibility.Always =>
+        true
+      case VersionCompatibility.Strict | VersionCompatibility.Default | VersionCompatibility.PackVer =>
+        // In PVP, any release can break source compatibility
+        previousVersion == currentVersion
+      case VersionCompatibility.SemVer | VersionCompatibility.EarlySemVer =>
+        val (currentMajor, currentMinor, currentPatch) = extractSemVerNumbers(currentVersion)
+        val (previousMajor, previousMinor, previousPatch) = extractSemVerNumbers(previousVersion)
+        currentMajor == previousMajor && {
+          if (currentMajor == 0)
+            currentMinor == previousMinor && currentPatch == previousPatch
+          else
+            currentMinor == previousMinor && currentPatch >= previousPatch
+        }
+      case VersionCompatibility.SemVerSpec =>
+        val (currentMajor, currentMinor, currentPatch) = extractSemVerNumbers(currentVersion)
+        val (previousMajor, previousMinor, previousPatch) = extractSemVerNumbers(previousVersion)
+        currentMajor == previousMajor &&
+          currentMinor == previousMinor &&
+          currentPatch >= previousPatch
     }
+
+  private def extractSemVerNumbers(versionString: String): (Int, Int, Int) = {
+    val version = Version(versionString)
+    if (!(version.items.size == 3 && version.items.forall(_.isInstanceOf[Version.Number]))) {
+      throw new MessageOnlyException(s"Version ${versionString} is not a semantic version. Set the correct versioning scheme via versionPolicyDependencySchemes.")
```

Comment:
Isn't that going to blow up if we depend on an M/RC of a library, even temporarily? Is this the desired behavior?

Reply:
> Isn't that going to blow up if we depend on an M/RC of a library, even temporarily?

Yes.

> Is this the desired behavior?

Yes, but now that I’m thinking more about it, I have a better idea. Let me push it.